src/protocols/secure_channel/CASESession.cpp

```diff
@@ -498,114 +461,103 @@ CHIP_ERROR CASESession::SendSigmaR2()
     VerifyOrExit(msg_salt.Alloc(saltlen), err = CHIP_ERROR_NO_MEMORY);
     VerifyOrExit(msg_rand.Alloc(kSigmaParamRandomNumberSize), err = CHIP_ERROR_NO_MEMORY);
 
-    // Step 1
     // Fill in the random value
     err = DRBG_get_bytes(msg_rand.Get(), kSigmaParamRandomNumberSize);
     SuccessOrExit(err);
 
-    // Step 3
-    // hardcoded to use a p256keypair
+    // Generate an ephemeral keypair
 #ifdef ENABLE_HSM_CASE_EPHERMAL_KEY
     mEphemeralKey.SetKeyId(CASE_EPHEMERAL_KEY);
 #endif
     err = mEphemeralKey.Initialize();
     SuccessOrExit(err);
 
-    // Step 4
+    // Generate a Shared Secret
     err = mEphemeralKey.ECDH_derive_secret(mRemotePubKey, mSharedSecret);
     SuccessOrExit(err);
 
-    err = ComputeIPK(mConnectionState.GetLocalKeyID(), mIPK, sizeof(mIPK));
-    SuccessOrExit(err);
-
-    // Step 5
     {
         MutableByteSpan saltSpan(msg_salt.Get(), saltlen);
-        err = ConstructSaltSigmaR2(ByteSpan(msg_rand.Get(), kSigmaParamRandomNumberSize), mEphemeralKey.Pubkey(), mIPK,
-                                   sizeof(mIPK), saltSpan);
+        err = ConstructSaltSigmaR2(ByteSpan(msg_rand.Get(), kSigmaParamRandomNumberSize), mEphemeralKey.Pubkey(),
+                                   ByteSpan(mIPK, sizeof(mIPK)), saltSpan);
         SuccessOrExit(err);
     }
 
     err = mHKDF.HKDF_SHA256(mSharedSecret, mSharedSecret.Length(), msg_salt.Get(), saltlen, kKDFSR2Info, kKDFInfoLength, sr2k,
                             kAEADKeySize);
     SuccessOrExit(err);
 
-    // Step 6
-    msg_r2_signed_len = static_cast<uint16_t>(sizeof(uint16_t) + mOpCredSet->GetDevOpCredLen(mTrustedRootId) +
-                                              kP256_PublicKey_Length * 2 + sizeof(uint64_t) * 3);
+    // Construct Sigma2 TBS Data
+    msg_r2_signed_len =
+        static_cast<uint16_t>(mOpCredSet->GetDevOpCredLen(mTrustedRootId) + kP256_PublicKey_Length * 2 + sizeof(uint64_t) * 3);
 
     VerifyOrExit(msg_R2_Signed.Alloc(msg_r2_signed_len), err = CHIP_ERROR_NO_MEMORY);
 
-    // Generate Sigma2 TBS Data
     {
         TLV::TLVWriter tlvWriter;
         TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
 
         tlvWriter.Init(msg_R2_Signed.Get(), msg_r2_signed_len);
         SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kResponderEphPubKey, mEphemeralKey.Pubkey(),
-                                               static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kNOC, mOpCredSet->GetDevOpCred(mTrustedRootId),
+        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(1), mOpCredSet->GetDevOpCred(mTrustedRootId),
```

Comment:
The spec calls for a `chip-certificate-array` format. When reading all the docs for `GetDevOpCred` and associated values, it's not clear at all that the format is chip-certificate-array. Is it the case?

Reply:
All certificates are stored in chip format inside OperationalCredentialSet class.

Reply:
Yes, they are stored as CHIP format, but what is the format output by `mOpCredSet->GetDevOpCred(mTrustedRootId)` ?

Reply:
This comment was resolved before without an answer.

Reply:
They are split into separate certificates before loading them onto the OperationalCredentialSet class. This is performed by the ExtractCertsFromCertArray method present in CHIPCertFromX509.cpp, which also verifies for kTLVType_Array